### ISSUE TITLE
Overload error should appear when defining the function, not when calling the function

### ISSUE THREAD

**Issue by thomasaarholt** [microsoft/pyright#5194]
**Describe the bug**
I have been typing some functions for a library and have had difficulty with the following case. I was trying to make mypy and pyright happy, and ended up accidentally committing code that makes both mypy and pyright happy with my definitions, but makes pyright complain when calling the function.

I have the following functions:
```python
# current version
from typing import overload, Literal
@overload
def sample(foo: int = 300, *, return_string: Literal[True] = True) -> str: ...


@overload
def sample(foo: int = 300, *, return_string: Literal[False] = False) -> int: ...


def sample(foo: int = 300, *, return_string: bool = True) -> str | int:
    if return_string:
        return "foo"
    else:
        return 1
```
pyright is happy with this, and I will argue that either it shouldn't be, or the following should be allowed. When I try to call the above function with only the first argument, I get an error:

```python
sample(foo=1000)
# No overloads for "sample" match the provided arguments
#   Argument types: (Literal[1000])
```

I understand the origin of the above error, even though the text itself is unclear: I haven't provided defaults for the overloaded definitions of `return_string`. Originally, I actually did:

```python
# previous version
@overload
def sample(
    foo: int = 300,
    *,
    return_string: Literal[True]
) -> str: ...

@overload
def sample(
    foo: int = 300,
    *,
    return_string: Literal[False]
) -> int: ...
```
and pyright was happy, but mypy wasn't (`Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]`), which I do believe is a bug on mypy's part. Then I experimented a bit, and landed on the version above, which didn't trigger any errors. It was only later that I noticed that when I tried to call the function in VSCode that I was getting the "no overloads" error from pyright.

I believe that pyright should complain sooner about the version above, preferably when one writes one version of the function with a default argument, and one without. This seems to be the crux of what the error is complaining about - although I may have misunderstood the error entirely, in which case I'm more than happy to be corrected!

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.
- Pylance vscode extension v2023.5.40

**Comment by erictraut** [comment]
I'm a bit confused by your post. If I understand you correctly, you said that you started with the following:

```python
from typing import overload, Literal

@overload
def sample(foo: int = 300, *, return_string: Literal[True]) -> str:
    ...

@overload
def sample(foo: int = 300, *, return_string: Literal[False]) -> int:
    ...

def sample(foo: int = 300, *, return_string: bool = True) -> str | int:
    if return_string:
        return "foo"
    else:
        return 1
```

And you said that pyright was happy with this but mypy was not. For me, this type checks without errors in both pyright and mypy. Of course, with this version, you will need to specify a keyword argument `return_string` that is either `True` or `False`. If you call it like `sample(foo=1000)`, you should receive a type error at the call site, and indeed you do with both mypy and pyright.

You said that you then changed the overload as follows:

```python
from typing import overload, Literal

@overload
def sample(foo: int = 300, *, return_string: Literal[True] = True) -> str:
    ...

@overload
def sample(foo: int = 300, *, return_string: Literal[False] = False) -> int:
    ...

def sample(foo: int = 300, *, return_string: bool = True) -> str | int:
    if return_string:
        return "foo"
    else:
        return 1
```

Now, you should be able to call `sample(foo=1000)` without error because the `return_string` parameter has a default argument. You imply above that pyright gives you an error at the call site here, but I don't see any such error with pyright or mypy. Mypy does report that there is an overlapping overload in this case. Arguably, pyright should as well. 

Please confirm that my understanding is correct because the behavior I'm observing doesn't match your description above.

**Comment by thomasaarholt** [comment]
Thanks for the quick reply, and apologies for the confusion, let me try to clear it up. You overall understand me correctly, and it's probably my own misunderstanding that its leading to my post's lack of clarity. One thing you misunderstood is that the bottom code block ("previous version") is what I started with and the top one ("current version") is the one I ended up with (but then was surprised by its behavior).

> Of course, with this version, you will need to specify a keyword argument return_string that is either True or False.

That is indeed what I notice - I have to include the keyword argument. I understand from your post that this is intended behavior, but I find it a bit surprising - this behavior seems to "not care" about the default argument passed in the final definition (`return_string: bool = True`) in the top ("current version") code block.

> You can eliminate this error by deleting the default argument in the second overload (i.e. delete the text = False).

Oh boy. So, I tried something similar:

```python
@overload
def sample(foo: int = 300, return_string: Literal[True] = True) -> str:
    ...

@overload
def sample(foo: int = 300, return_string: Literal[False]) -> int:
    ...
```
That is, without the keyword-only argument asterix (`*`). This gave me `Non-default argument follows default argument` in pyright - which made sense, and I didn't think to try it with it (the asterix). Thank you, this solves the "incompatibility" that I was having between mypy and pyright.

I'll leave the issue open in case you want to respond to it, but the issue can be closed.

**Comment by thomasaarholt** [comment]
I guess this part of the [pyright overload resolution documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads) is relevant here?
> 4. If more than one overload remains, the “winner” is chosen based on the order in which the overloads are declared. In general, the first remaining overload is the “winner”.

I can't seem to find any pyright or pep484 documentation explaining the usage of asterix here. It was by chance that the original `sample` function that I was trying to overload had a keyword-only asterix present. [Here is the original function](https://github.com/pymc-devs/pymc/blob/fbc62d55f12b15c1dda81c6cc51a1b4790ac3eb0/pymc/sampling/mcmc.py#L391) in the pymc library.

**Comment by erictraut** [comment]
> This behavior seems to "not care" about the default argument passed in the final definition

The "final definition" is referred to as the "implementation" of the function, to distinguish it from the "overloads".

When a type checker evaluates a call expression and the target function has overloads, only overload signatures are considered. The signature of the implementation is completely ignored. This is the behavior dictated by PEP 484, which introduced the notion of `@overload`.

Here's the relevant part of PEP 484:
> The `@overload`-decorated definitions are for the benefit of the type checker only, since they will be overwritten by the non-`@overload`-decorated definition, while the latter is used at runtime but should be ignored by a type checker.

Incidentally, if you are defining an overloaded function or method in a type stub, an abstract class or a protocol class, you can completely omit the implementation of an overloaded function because the type checker looks only at the `@overload` signatures.

**Comment by thomasaarholt** [comment]
That makes sense! I hadn't really realized that the implementation was completely ignored by the type checker. This could possibly be made clearer in the pyright documentation. I guess it's an implication of the resolution order rules (rule 6 starts with `"If no overloads remain"`), but could be made explicit there.

**Comment by thomasaarholt** [comment]
Thank you very much for your very clear comments.
